### DISCUSSION TITLE
How to use getCharactersAt

### DISCUSSION THREAD

**Discussion by sbernardotti** [Annoraaq/grid-engine#361] [Q&A] [answered]
Hi everyone, I need to check if there is a player in determined position

What I found is the funcion getCharactersAt which receives 2 parameters: position (x, y) and layer (string).

The problem is that I always receive an empty array. Here is the test code:

```
//////
//////
this.gridEngine.addCharacter({
      id: "player",
      container: playerContainer,
      speed: 4.1,
      charLayer:"charlayer"
});

this.gridEngine.setPosition("player", { x: 50, y: 50 });

const checkFunc = this.gridEngine.getCharactersAt({ x: 50, y: 50 }, "charlayer");

// Expected: ["player"]
// Returns: []

//////
//////
```

Thank you!

**Comment by Annoraaq** [comment, marked as answer]
I think in your case the problem is that you omit the `charLayer` argument of `setPosition`. Therefore it defaults to `undefined`. So you position the character to charLayer `undefined`. 
Change it to
```this.gridEngine.setPosition("player", { x: 50, y: 50 }, "charLayer");```
and it should work.

Now that I see it, I must say that I can understand the confusion here. It is not very intuitive that the char layer is also changed if you do not pass it.

I think a reason for this was compatibility for tilemaps without char layers. If no char layers are specified in a tile map, the uppermost layer that does not have the `ge_alwaysTop` property is selected as the default char layer, called `undefined`.
And it is an optional parameter because of backwards compatibility with GridEngine versions that did not support char layers.

One alternative would be to change setPosition in a way, that it does not change the char layer at all, if it is set to undefined.
But since this is a backwards compatibility breaking change, I will postpone it to version 3 of GridEngine.

However, I will update the docs, so that it is explicit that the layer `undefined` is set, if the parameter is omitted.
